src/app/WriteClient.h

```diff
@@ -0,0 +1,141 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines read client for a CHIP Interaction Data model
+ *
+ */
+
+#pragma once
+
+#include <app/AttributePathParams.h>
+#include <app/InteractionModelDelegate.h>
+#include <app/MessageDef/AttributeDataList.h>
+#include <app/MessageDef/AttributeStatusElement.h>
+#include <app/MessageDef/WriteRequest.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPTLVDebug.hpp>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace app {
+/**
+ *  @class WriteClient
+ *
+ *  @brief The read client represents the initiator side of a Write Interaction, and is responsible
+ *  for generating one Write Request for a particular set of attributes, and handling the Write response.
+ *
+ */
+class WriteClient : public Messaging::ExchangeDelegate
+{
+public:
+    /**
+     *  Shutdown the WriteClient. This terminates this instance
+     *  of the object and releases all held resources.
+     *
+     */
+
+    void Shutdown();
+
+    /**
+     * Finalize Write Request Message TLV Builder
+     *
+     *
+     */
+    CHIP_ERROR FinalizeMessage(System::PacketBufferHandle & aPacket);
+
+    /**
+     *  Send a Write Request.  There can be one Write Request outstanding on a given WriteClient.
+     *  If SendWriteRequest returns success, no more Write Requests can be sent on this WriteClient
+     *  until the corresponding InteractionModelDelegate::WriteResponseProcessed or InteractionModelDelegate::WriteResponseError
+     *  call happens with guarantee.
+     */
+    CHIP_ERROR SendWriteRequest(NodeId aNodeId, Transport::AdminId aAdminId, SecureSessionHandle * apSecureSession);
+
+    CHIP_ERROR PrepareAttribute(const AttributePathParams & attributePathParams);
+    CHIP_ERROR FinishAttribute();
+    TLV::TLVWriter * GetAttributeDataElementTLVWriter();
+
+private:
+    friend class TestWriteInteraction;
```

Comment:
can we implement things without friend classes?

Friend classes signal very strong coupling and are often not needed - this seems to try to protect developers from themselves - why?

Reply:
one reason is that when testing received message for client, void WriteClient::OnMessageReceived override OnMessageReceived in exchange context, which don't return any error, so I have to test internal private function, CHIP_ERROR ProcessWriteResponseMessage, then i need friend class TestWriteInteraction. I would create the ticket to figure out how to better decouple this for testing.

Reply:
https://github.com/project-chip/connectedhomeip/issues/7754